spec/src/main/asciidoc/servlet-spec-body.adoc

```diff
@@ -1303,6 +1303,145 @@ the header value to an `int`, a `NumberFormatException` is thrown. If
+
+HTTP/2:: The URI path is the `:path` pseudo header as defined by [RFC 7540](https://datatracker.ietf.org/doc/html/rfc7540#section-8.1.2.3) and is passed unchanged to stage 2.
+
+HTTP/3:: The URI path is the `:path` pseudo header as currently defined by the [draft RFC](https://datatracker.ietf.org/doc/html/draft-ietf-quic-http-34).
+
+Other protocols:: Containers may support other protocols. Containers should extract an appropriate URI path for the request from the protocol and pass it to stage 2.
+
+==== URI Path Canonicalization
+
+Servlet containers may implement the standard Servlet URI path canonicalization in any manner they see fit as long as the end result is identical to the end result of the process described here. Servlet containers may provide container specific configuration options to vary the standard canonicalization process. Any such variations may have security implications and both Servlet container implementors and users are advised to be sure that they understand the implications of any such container specific canonicalization options.
+
+. **Separation of path and query.**
++
+The URI is split by the first occurrence of any '"?"' character to path and query.  The query is preserved for later handling and the following steps applied to the path.
+
+. **Discard fragment.**
++
+The path is split by the first occurrence of any `"\#"` character. The `"#"` and following fragment are discarded and the path is replaced with the character sequence preceding the `"#"` character.
+
+. **Split path into segments.**
++
+The path is split into multiple segments divided by the `"/"` character. A leading `"/"` does not indicate an initial empty segment.
+
+. **Remove path parameters.**
++
+Any segment containing the `";"` character is split at the first occurrence of `";"`. The segment is replaced by the character sequence preceding the `";"`. The characters following the `";"` are considered path parameters and may be preserved by the container for later decoding and/or processing (eg `jsessionid`).
+
+. **Decode characters.**
++
+Characters that are encoded in `%nn` form are decoded in each segment. The resulting octet sequences is treated as UTF-8 and converted to a character sequence that replaces the segment.
+
+. **Remove Empty Segments.**
++
+Empty segments are removed.  Containers may be configured to retain empty segments.
+
+. **Remove dot-segments.**
++
+All segments that are exactly `"."` are removed from the segment series. Segments that are exactly `".."` and that are preceded by a non `".."` segment are removed together with the preceding segment.     This normalization differs from RFC3986 in that segments with `%` encoded dots and/or parameters may be treated as dot segments.
```

Comment:
I've seen this view (`/%2e%2e/` != `/../`) expressed in multiple places but I can find text in RFC 3986 to support it. Section 2.3 seems to say exactly the opposite. What am I missing?

Reply:
Ah very VERY good point.  I think I've tricked myself into thinking that because the RFC does allow reserved chars like `/` to be encoded to exclude them from normalization.... so I think I had assumed that `.` could also be encoded to exclude it from normalization.   But reading it again now, you're right, that `.` encoding must be decoded before normilization.       So we are mostly doing the right thing after all and I need not fret so much about it.      We still have a few differences, but not huge ones. I'll update text.